api/transport/request.go

```diff
@@ -62,6 +62,10 @@ type Request struct {
 
 	// Request payload.
 	Body io.Reader
+
+	// TransportHeaders are other transport headers that the transport may expose
+	// to the encodings.
+	TransportHeaders map[string]string
```

Comment:
Thanks, I'll add that approach to the list of alternatives in the description.

Pros: simple, doesn't require new APIs
Cons: effectively reserves a value in the application header namespace (right now we allow any application header and we namespace them using `RPC-Header-` prefix iirc)